Instance options (#119)
meilisearch/specifications: text/0119-instance-options.md

- Title: Instance options

> - Start Date: 01.03.2022

Reviewer: I've gotten used to using the YYYY-MM-DD format over time 😊

Example with the snapshort creation:
- `export MEIL_SCHEDULE_SNAPSHOT=yes` means the snapshot creation is enabled.
- `export MEIL_SCHEDULE_SNAPSHOT=off` means the snapshot creation is disabled.
- No variable set means the snapshot creation is disabled.

### Error bahavior

1. Some configuration options must specify a value. Using a command-line option (that requires a value) or an environment variable without specifying a value will throw an error and interrupt the launch process.

Example:

❌ Wrong

```bash
./meilisearch --db-path

error: The argument '--db-path <DB_PATH>' requires a value but none was supplied
```

✅ Correct

```bash
./meilisearch --db-path ./meilifiles
```

2. Some command-line options take an implicit boolean as a value. In this case, the users should not set any value when using the option.

❌ Wrong

```bash
./meilisearch --schedule-snapshot yes

error: Found argument 'yes' which wasn't expected, or isn't valid in this context
```

✅ Correct

```bash
./meilisearch --schedule-snapshot
```

The expected behavior of each flag is described in the list above.

### Exhaustive list of options

> - [Database path](#database-path)
> - [Environment](#environment)
> - [HTTP address & port binding](#http-address-port-binding)
> - [Master key](#master-key)
> - [Disable analytics](#disable-analytics)
> - [Dumps](#dumps-destination)
> - [Dumps destination](#dumps-destination)
> - [Import dump](#import-dump)
> - [Log level](#log-level)
> - [Max index size](#max-index-size)
> - [Max TASK_DB size](#max-task-db-size)
> - [Payload limit size](#payload-limit-size)
> - [Snapshots](#schedule-snapshot-creation):
> - [Schedule snapshot creation](#schedule-snapshot-creation)
> - [Snapshot destination](#snapshot-destination)
> - [Snapshot interval](#snapshot-interval)
> - [Import snapshot](#import-snapshot)
> - [Ignore missing snapshot](#ignore-missing-snapshot)
> - [Ignore snapshot if DB exists](#ignore-snapshot-if-db-exists)
> - [SSL configuration](#ssl-authentication-path):
> - [SSL authentication path](#ssl-authentication-path)
> - [SSL certificates path](#ssl-certificates-path)
> - [SSL key path](#ssl-key-path)
> - [SSL OCSP path](#ssl-ocsp-path)
> - [SSL require auth](#ssl-require-auth)
> - [SSL resumption](#ssl-resumption)
> - [SSL tickets](#ssl-tickets)
> - [Database path](#database-path)
> - [Environment](#environment)
> - [HTTP address & port binding](#http-address-port-binding)
> - [Master key](#master-key)
> - [Disable analytics](#disable-analytics)
> - [Dumps](#dumps-destination)
> - [Dumps destination](#dumps-destination)
> - [Import dump](#import-dump)
> - [Log level](#log-level)
> - [Max index size](#max-index-size)
> - [Max TASK_DB size](#max-task-db-size)
> - [Payload limit size](#payload-limit-size)
> - [Snapshots](#schedule-snapshot-creation)
> - [Schedule snapshot creation](#schedule-snapshot-creation)
> - [Snapshot destination](#snapshot-destination)
> - [Snapshot interval](#snapshot-interval)
> - [Import snapshot](#import-snapshot)
> - [Ignore missing snapshot](#ignore-missing-snapshot)
> - [Ignore snapshot if DB exists](#ignore-snapshot-if-db-exists)
> - [Memory usage](#memory-usage)
> - [Indexing jobs](#indexing-jobs)
> - [SSL configuration](#ssl-authentication-path)
> - [SSL authentication path](#ssl-authentication-path)
> - [SSL certificates path](#ssl-certificates-path)
> - [SSL key path](#ssl-key-path)
> - [SSL OCSP path](#ssl-ocsp-path)
> - [SSL require auth](#ssl-require-auth)
> - [SSL resumption](#ssl-resumption)
> - [SSL tickets](#ssl-tickets)

Reviewer: ```suggestion
- [Database path](#database-path)
- [Environment](#environment)
- [HTTP address & port binding](#http-address-port-binding)
- [Master key](#master-key)
- [Disable analytics](#disable-analytics)
- [Dumps](#dumps-destination)
- [Dumps destination](#dumps-destination)
- [Import dump](#import-dump)
- [Log level](#log-level)
- [Max index size](#max-index-size)
- [Max TASK_DB size](#max-task-db-size)
- [Payload limit size](#payload-limit-size)
- [Snapshots](#schedule-snapshot-creation)
- [Schedule snapshot creation](#schedule-snapshot-creation)
- [Snapshot destination](#snapshot-destination)
- [Snapshot interval](#snapshot-interval)
- [Import snapshot](#import-snapshot)
- [Ignore missing snapshot](#ignore-missing-snapshot)
- [Ignore snapshot if DB exists](#ignore-snapshot-if-db-exists)
- [Memory usage](#memory-usage)
- [Indexing jobs](#indexing-jobs)
- [SSL configuration](#ssl-authentication-path)
- [SSL authentication path](#ssl-authentication-path)
- [SSL certificates path](#ssl-certificates-path)
- [SSL key path](#ssl-key-path)
- [SSL OCSP path](#ssl-ocsp-path)
- [SSL require auth](#ssl-require-auth)
- [SSL resumption](#ssl-resumption)
- [SSL tickets](#ssl-tickets)
```

Reviewer: Removing duplicated value here. I've kept the list with the newly added options.

#### HTTP address & port binding

**Environment variable**: `MEILI_HTTP_ADDR`
**CLI option**: `--http-addr`
**Default value**: `"127.0.0.1:7700"`
**Expected value**: an HTTP address and port

Sets the HTTP address and port Meilisearch will use.

#### Master key

**Environment variable**: `MEILI_MASTER_KEY`
**CLI option**: `--master-key`
**Default value**: `None`
**Expected value**: an alphanumeric string

Sets the instance's master key, automatically protecting all routes except [`GET /health`](/reference/api/health.md). This means you will need an API key to access endpoints such as `POST /search` and `GET /documents`.

You must supply an alphanumeric string when using this option.

Providing a master key is mandatory when `--env` is set to `production`; if none is given, then Meilisearch will throw an error and refuse to launch.

If no master key is provided in a `development` environment, all routes will be unprotected and publicly accessible.

#### Disable analytics

**Environment variable**: `MEILI_NO_ANALYTICS`
**CLI option**: `--no-analytics`
**Default**: Enabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

> Deactivates Meilisearch's built-in telemetry when enabled.

Reviewer: ```suggestion
Deactivates Meilisearch's built-in telemetry collect when enabled.
```

### Dumps destination

**Environment variable**: `MEILI_DUMPS_DIR`
**CLI option**: `--dumps-dir`
**Default value**: `dumps/`
**Expected value**: a filepath pointing to a valid directory

> Sets the directory where Meilisearch will create dump files.

Reviewer: ```suggestion
Sets the directory where Meilisearch will create dump files. If the directory does not exist when a dump is generated it will be created.
```

> ### Import dump

Reviewer: Options `--ignore-missing-dump` and `--ignore-dump-if-db-exists` are gone in the wild 👻

They are described [here](https://github.com/meilisearch/specifications/blob/develop/text/0105-dumps-api.md) 

What do you think about adding a helper like this at the end of each option that is related to the dump feature? 🤓

> See [Dumps API](0105-dumps-api.md)

The `task` database handles pending tasks. This is different from the `index` database, which only stores processed data.

#### Payload limit size

**Environment variable**: `MEILI_HTTP_PAYLOAD_SIZE_LIMIT`
**CLI option**: `--http-payload-size-limit`
**Default value**: `104857600` (~100MB)
**Expected value**: an integer (`104857600`) or a human readable size (`100Mb`)

Sets the maximum size of accepted payloads. Value must be given in bytes or explicitly stating a base unit. For example, the default value can be written as `107374182400`, `'107.7Gb'`, or `'107374 Mb'`.

#### Schedule snapshot creation

**Environment variable**: `MEILI_SCHEDULE_SNAPSHOT`
**CLI option**: `--schedule-snapshot`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Activates scheduled snapshots when enabled. Snapshots are disabled by default.

#### Snapshot destination

**Environment variable**: `MEILI_SNAPSHOT_DIR`
**CLI option**: `--snapshot-dir`
**Default value**: `snapshots/`
**Expected value**: a filepath pointing to a valid directory

> Sets the directory where Meilisearch will store snapshots.

Reviewer: ```suggestion
Sets the directory where Meilisearch will store snapshots. If the directory does not exist when a snapshot is generated it will be created.
```

This command will throw an error if `--import-snapshot` is not defined.

#### Ignore snapshot if DB exists

**Environment variable**: N/A
**CLI option**: `--ignore-snapshot-if-db-exists`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Prevents a Meilisearch instance with an existing database from throwing an error when using `--import-snapshot`. Instead, the snapshot will be ignored and Meilisearch will launch using the existing database.

This command will throw an error if `--import-snapshot` is not defined.

#### Memory usage

**Environment variable**: `MEILI_MAX_MEMORY`
**CLI option**: `--max-memory`
**Default value**: 2/3 of the available RAM of the machine
**Expected value**: an integer (`104857600`) or a human readable size (`100Mb`)

Set the maximum size of the RAM used by Meilisearch. By default, Meilisearch adapts its behavior to make the indexation use a maximum two thirds of available resources.

Value must be given in bytes or explicitly stating a base unit. For example, the default value can be written as `107374182400`, `'107.7Gb'`, or `'107374 Mb'`.

⚠️ WARNINGS

> - We do not recommend setting the full size of your machine. For example, when running Meilisearch on a machine with 4GB of RAM, do not set this options to `4Gb`

Reviewer: ```suggestion
- We do not recommend setting the full RAM size of your machine. For example, when running Meilisearch on a machine with 4GB of RAM, do not set this options to `4Gb`
```

- This command-line option will not perfectly ensure the RAM usage, but will help you manage multiple Meilisearch engines on the same machine (for example, using Kubernetes). The core team cannot guarantee the exact usage of the RAM.
- If the number set is higher than the real available RAM in the machine, we cannot prevent Meilisearch from crashing.

#### Indexing jobs

**Environment variable**: `MEILI_INDEXING_JOBS`
**CLI option**: `--indexing-jobs`
**Default value**: half of the available core of the machine
**Expected value**: an integer

Sets the number of jobs available during the indexation.

By default, in machines with multi-core processors, the indexer avoids using more than half of the available processing units. For example, if your machine has twelve cores, the indexer will try to use six of them at most. This ensures Meilisearch is always ready to perform searches, even while you are updating an index.

> Obviously, multi-threading is not possible in machines with only one processor core.

Reviewer: Not sure if it's useful to add that precision.

- Title: Instance options
- Start Date: 2022-03-01

# Instance options

## 1. Summary

The instance options let the users configure Meilisearch when launching the search engine using
- environment variables
- command-line options

An example when setting an environment variable to change the folder path where the Meilisearch data are stored:

```bash
export MEILI_DB_PATH=./meilifiles
./meilisearch
```

Same behavior using the command-line option:

```bash
./meilisearch --db-path ./meilifiles
```

## 2. Motivation

When Meilisearch is launched, the default configuration may not meet the specific needs of users. Meilisearch exposes configurable options to allow users to fine-tune the behavior of the search engine.

## 3. Functional Specification

The users can configure Meilisearch when launching the search engine using
- **environment variables**. Ex: `--db-path`
- **command-line options or CLI options**. Ex: `MEILI_DB_PATH`

There are 2 categories of CLI (command-line interface) options:

> - the ones that expect a value. Ex: `--db-path`.

Reviewer 2: ```suggestion
- the ones that expect a value. Ex: `--db-path "./my_path"`.
```

> - the ones that does expect any value, called also "flags". Ex: `--no-analytics`. Their implicit values are booleans.

Reviewer 3: Shouldn't this be `don’t expect values`?

Author: Indeed!

> - the ones that does expect any value, called also "flags". Ex: `--no-analytics`. Their implicit values are booleans.

Author: ```suggestion
- the ones that don't expect any value, called also "flags". Ex: `--no-analytics`. Their implicit values are booleans.
```

### 3.1. Some specific behaviors

#### 3.1.1. Priority between CLI options and environment variables

Command-line options take precedence over environment variables. If the same configuration option is specified both as a command-line option and as an environment variable, Meilisearch will use the command-line option and its respective value.

#### 3.1.2. Flags: accepted values for the corresponding environment variable

The options that do not expect any value when using the command-line option accepts the following value when using the corresponding environment variable: `n`, `no`, `f`, `false`, `off`, and `0` as `false`. An absent environment variable will also be considered as `false`. Everything else is considered `true`.

Example with the snapshort creation:

> - `export MEIL_SCHEDULE_SNAPSHOT=yes` means the snapshot creation is enabled.
> - `export MEIL_SCHEDULE_SNAPSHOT=off` means the snapshot creation is disabled.

Reviewer 4: ```suggestion
- `export MEILI_SCHEDULE_SNAPSHOT=yes` means the snapshot creation is enabled.
- `export MEILI_SCHEDULE_SNAPSHOT=off` means the snapshot creation is disabled.
```

#### Environment

**Environment variable**: `MEILI_ENV`
**CLI option**: `--env`
**Default value**: `development`
**Expected value**: `production` or `development`

Configures the instance's environment. Value must be either `production` or `development`.

`production`:

- Setting a master key is **mandatory**
- The search preview interface is disabled

`development`:

- Setting a master key is **optional**
- Search preview is enabled

#### HTTP address & port binding

**Environment variable**: `MEILI_HTTP_ADDR`
**CLI option**: `--http-addr`
**Default value**: `"127.0.0.1:7700"`
**Expected value**: an HTTP address and port

Sets the HTTP address and port Meilisearch will use.

#### Master key

**Environment variable**: `MEILI_MASTER_KEY`
**CLI option**: `--master-key`
**Default value**: `None`
**Expected value**: an alphanumeric string

Sets the instance's master key, automatically protecting all routes except [`GET /health`](/reference/api/health.md). This means you will need an API key to access endpoints such as `POST /search` and `GET /documents`.

You must supply an alphanumeric string when using this option.

> Providing a master key is mandatory when `--env` is set to `production`; if none is given, then Meilisearch will throw an error and refuse to launch.

Reviewer 5: Do we have some kind of validation in this option? Like: the `master-key` should have 10 characters at least to work?

Author: Nope, no validation, but can be considered in the future (it will be annoying for the integration team that currently run all the test with `masterKey` as master key 😂)

Reviewer 5: YEAAAH! hahah 

Could be only required for the `production` env by the way :)

If no master key is provided in a `development` environment, all routes will be unprotected and publicly accessible.

#### Disable analytics

**Environment variable**: `MEILI_NO_ANALYTICS`
**CLI option**: `--no-analytics`
**Default**: Enabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Deactivates Meilisearch's built-in telemetry collect when enabled.

> ### Dumps destination

Reviewer: ```suggestion
#### Dumps destination
```

**Environment variable**: N/A
**CLI option**: `--ignore-missing-snapshot`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Prevents a Meilisearch instance from throwing an error when `--import-snapshot` does not point to a valid snapshot file.

This command will throw an error if `--import-snapshot` is not defined.

#### 3.3.19. Ignore snapshot if DB exists

**Environment variable**: N/A
**CLI option**: `--ignore-snapshot-if-db-exists`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Prevents a Meilisearch instance with an existing database from throwing an error when using `--import-snapshot`. Instead, the snapshot will be ignored and Meilisearch will launch using the existing database.

This command will throw an error if `--import-snapshot` is not defined.

#### 3.3.20. Max memory usage when indexing

**Environment variable**: `MEILI_MAX_INDEXING_MEMORY`
**CLI option**: `--max-indexing-memory`
**Default value**: 2/3 of the available RAM of the machine
**Expected value**: an integer (`104857600`) or a human readable size (`100Mb`)

> Set the maximum size of the RAM used by Meilisearch when indexing. By default, Meilisearch adapts its behavior to make the indexation use a maximum two thirds of available resources.

Reviewer 2: ```suggestion
Set the maximum size of the RAM used by Meilisearch when indexing. By default, Meilisearch adapts its behavior to make the indexation use at maximum two-thirds of the available resources.
```

Set the maximum size of the RAM used by Meilisearch. By default, Meilisearch adapts its behavior to make the indexation use a maximum two thirds of available resources.

Value must be given in bytes or explicitly stating a base unit. For example, the default value can be written as `107374182400`, `'107.7Gb'`, or `'107374 Mb'`.

⚠️ WARNINGS
- We do not recommend setting the full RAM size of your machine. For example, when running Meilisearch on a machine with 4GB of RAM, do not set this options to `4Gb`

> - This command-line option will not perfectly ensure the RAM usage, but will help you manage multiple Meilisearch engines on the same machine (for example, using Kubernetes). The core team cannot guarantee the exact usage of the RAM.

Reviewer 5: I'm not sure about mentioning the core team here. Sounds like an entity that will handle that for the user 😄. What about: `The search engine cannot guarantee the exact usage of the RAM` 🤓

> - This command-line option will not perfectly ensure the RAM usage, but will help you manage multiple Meilisearch engines on the same machine (for example, using Kubernetes). The search engine cannot guarantee the exact usage of the RAM.

Reviewer 2: ```suggestion
- This command-line option does not perfectly ensure the RAM usage but helps you manage multiple Meilisearch engines on the same machine (for example, using Kubernetes). The search engine cannot guarantee the exact usage of the RAM.
```

> - If the number set is higher than the real available RAM in the machine, we cannot prevent Meilisearch from crashing.

Reviewer 4: Max memory isn't representative of what the variable does, this memory limit is only for the indexing and doesn't take into account searches and other meilisearch processes including task store.
`--indexing-memory` would be clearer and would be consistent with `--indexing-jobs`.

Reviewer 2: I agree with `--indexing-memory` or even, but it might become to long, `--indexing-max-memory`. Which I think is the most intuitive to understand.

#### 3.3.21. Indexing jobs

**Environment variable**: `MEILI_INDEXING_JOBS`

> **CLI option**: `--indexing-jobs`

Reviewer 2: It is not very clear that this parameter is requiring a number of processors. A job for me is more a number of tasks, not correlated with any hardware.

Im not the best in hardware naming but these are more understandable for me: 
- `cores`
- `processors`
- `threads`

In elasticseach the setting is called [`processors`](https://octoperf.com/blog/2018/09/21/optimizing-elasticsearch/#detected-processors).

I would suggest something in the line of `indexing-processors`. 
Or but it might become too long `--indexing-number-processor` (not sure about `number`).

Reviewer 4: only the term `threads` would be really accurate I think.

Author: Often the users in GitHub issues and discussions talk about thread as well

Reviewer 2: `--indexing-threads` or `--indexing-number-threads` is also good for me!

Reviewer 2: @gmourier if I set it to `5` does Meilisearch absolutely tries to use the 5 threads or does Meilisearch consider using them depending on the workload?

Reviewer 6: No, it does not necessarily use all cores, --max-indexing-threads is a good option

**Default value**: half of the available core of the machine
**Expected value**: an integer

> Sets the number of jobs available during the indexation.

Reviewer 2: Following my previous comment
```suggestion
Sets the number of processors available during the indexation.
```

Reviewer 4: ```suggestion
Sets the number of threads available during the indexation.
```

> **Default value**: half of the available core of the machine

Reviewer 2: ```suggestion
**Default value**: half of the available cores of the machine
```

> **Default value**: half of the available cores of the machine

Author: ```suggestion
**Default value**: half of the available threads of the machine
```

> Sets the maximum number of threads available during the indexation.

Reviewer 2: ```suggestion
Sets the maximum number of threads that Meilisearch can access during the indexation.
```

or 

```suggestion
Sets the maximum number of threads available to use during the indexation.
```

> By default, in machines with multi-core processors, the indexer avoids using more than half of the available processing units. For example, if your machine has twelve cores, the indexer will try to use six of them at most. This ensures Meilisearch is always ready to perform searches, even while you are updating an index.

Author: ```suggestion
By default, in machines with multi-core processors, the indexer avoids using more than half of the available processing units. For example, if your machine has twelve cores, the indexer will try to use six of them at most. This ensures Meilisearch is always ready to perform searches, even while you are updating an index.
Users have to be aware that requiring the totality of the threads for the indexation can lead to a huge impact on the search experience.
```

**Environment variable**: `MEILI_SSL_OCSP_PATH`
**CLI option**: `--ssl-ocsp-path`
**Default value**: `None`
**Expected value**: a filepath pointing to a valid OCSP certificate

Sets the server's OCSP file. *Optional*

Reads DER-encoded OCSP response from OCSPFILE and staple to certificate.

#### SSL require auth

**Environment variable**: `MEILI_SSL_REQUIRE_AUTH`
**CLI option**: `--ssl-require-auth`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Makes SSL authentication mandatory.

Sends a fatal alert if the client does not complete client authentication.

#### SSL resumption

**Environment variable**: `MEILI_SSL_RESUMPTION`
**CLI option**: `--ssl-resumption`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Activates SSL session resumption.

#### SSL tickets

**Environment variable**: `MEILI_SSL_TICKETS`
**CLI option**: `--ssl-tickets`
**Default**: Disabled

⚠️ This command-line option does not take any values. Assigning a value will throw an error.

Activates SSL tickets.

## 4. Technical Aspects

N/A

## 5. Future Possibilities

> - Redo the command-line to create a more interactive CLI

Reviewer 5: Do we have a autocomplete to the options (like `git` does)?

<img width="313" alt="image" src="https://user-images.githubusercontent.com/4116980/156416391-f0b0c4fc-8fb6-4124-b5bb-28b305b3c39c.png">

Author: Not as far as I know, can be a future possibility :)

Reviewer 5: Cool, yeah that's the idea, could be helpful to add in the `Future Possibilities`
